nengo/synapses.py

```diff
@@ -221,123 +197,123 @@ def evaluate(self, frequencies):
+            self.D = D
+            self.X = X
 
         def __call__(self, t, signal):
-            raise NotImplementedError("Step functions must implement __call__")
-
-    class NoDen(Step):
-        """An LTI step function for transfer functions with no denominator.
-
-        This step function should be much faster than the equivalent general
-        step function.
-        """
-        def __init__(self, num, den, output):
-            if len(den) > 0:
-                raise ValidationError("'den' must be empty (got length %d)"
-                                      % len(den), attr='den', obj=self)
-            super().__init__(num, den, output)
-            self.b = num[0]
+            raise NotImplementedError("Step object must implement __call__")
+
+        @classmethod
+        def check(cls, A, B, C, D, X):
+            if A.shape[0] == 0:
+                return X.shape[0] == B.shape[0] == 0 and D.shape[0] == 1
+            else:
+                return (A.shape[0] == A.shape[1] == B.shape[0] == C.shape[1]
+                        == X.shape[0]
+                        and A.shape[0] == X.shape[0]
+                        and D.shape[0] == B.shape[1] == 1)
+
+    class NoX(Step):
+        """Step for transfer function with no state and single numerator."""
+        def __init__(self, A, B, C, D, X):
+            super().__init__(A, B, C, D, X)
+            self.d = D.item()
 
         def __call__(self, t, signal):
-            self.output[...] = self.b * signal
-            return self.output
+            return self.d * signal
 
-    class Simple(Step):
-        """An LTI step function for transfer functions with one num and den.
+        @classmethod
+        def check(cls, A, B, C, D, X):
+            return super(LinearFilter.NoX, cls).check(A, B, C, D, X) and (
+                len(A) == 0)
 
-        This step function should be much faster than the equivalent general
-        step function.
-        """
-        def __init__(self, num, den, output, y0=None):
-            if len(num) != 1:
-                raise ValidationError("'num' must be length 1 (got %d)"
-                                      % len(num), attr='num', obj=self)
-            if len(den) != 1:
-                raise ValidationError("'den' must be length 1 (got %d)"
-                                      % len(den), attr='den', obj=self)
-
-            super().__init__(num, den, output)
-            self.b = num[0]
-            self.a = den[0]
-            if y0 is not None:
-                self.output[...] = y0
+    class OneX(Step):
+        """Step function for transfer functions with one state element."""
+        def __init__(self, A, B, C, D, X):
+            super().__init__(A, B, C, D, X)
+            self.a = A.item()
+            self.b = C.item()
 
         def __call__(self, t, signal):
-            self.output *= -self.a
-            self.output += self.b * signal
-            return self.output
+            self.X *= -self.a
```

Comment:
Is this being unit-tested? According to the coverage report it isn't, but not sure how flaky that is. The reason I figured was because I think this should be `self.X *= self.a` without the minus sign. Since the linear system is of the form `x = Ax + Bu`.

Reply:
You're right, it's not getting hit. That minus was left over from the transfer function implementation. Could be part of the reason why I was seeing a slowdown for Lowpass synapses.

I've added a commit now that I think fixes it.